Be sure to print the desired output. You will not receive credit for problems that do not print the desired output.

Many of these exercises will require the use of function in numpy that were not covered in class. You'll probably need to look up the documentation on how to use them. I always recommend checking the official docs at https://docs.scipy.org/doc/numpy/ before reading other sites

In [3]:
import numpy as np
np.random.seed(1)

Exercise 0: (done for you as an example)

Task: Make an array `a` with values 0 to 9

Desired output: `[0 1 2 3 4 5 6 7 8 9]`

In [29]:
a = np.arange(10)
print(a)

[0 1 2 3 4 5 6 7 8 9]


Exercise 1:

Task: Create an array `b` of 10 random integers selected between 0-99

Desired output: `[37 12 72  9 75  5 79 64 16  1]` of course yours might be different

In [30]:
b = np.random.randint(0, 99, 10)

Exercise 2:

Task: reverse the elements in `b`. Hint: Try slicing the array, but backwards

Desired output: `[ 1 16 64 79  5 75  9 72 12 37]` yours will be different

In [31]:
b[::-1]

array([ 1, 16, 64, 79,  5, 75,  9, 72, 12, 37])

Exercise 3:

Task: Create an array `c` of 10 random values selected from a normal distribution centered at 100 with sd = 15, rounded to 1 decimal place.

Desired output: `[ 92.1  83.9 113.   65.5 126.2  88.6 104.8  96.3 121.9  69.1]` of course your values may be different

In [32]:
c = np.round(np.random.normal(100, 15, 10), decimals=1)
c

array([ 92.1,  83.9, 113. ,  65.5, 126.2,  88.6, 104.8,  96.3, 121.9,
        69.1])

Exercise 4:

Task: sort all the values in c from least to greatest.

Desired output: `[ 65.5  69.1  83.9  88.6  92.1  96.3 104.8 113.  121.9 126.2]`

In [35]:
c.sort()
c

array([ 65.5,  69.1,  83.9,  88.6,  92.1,  96.3, 104.8, 113. , 121.9,
       126.2])

Exercise 5:

Task: Reshape `c` so that it is a  2 x 5 array. Save the result as `d`

Desired output: 

`[[ 65.5  69.1  83.9  88.6  92.1]
 [ 96.3 104.8 113.  121.9 126.2]]`

In [38]:
d = c.reshape(2,5)
d

array([[ 65.5,  69.1,  83.9,  88.6,  92.1],
       [ 96.3, 104.8, 113. , 121.9, 126.2]])

Exercise 6:

Task: Return the row sum and column sums of `d`

Desired output: 

`[399.2 562.2]
[161.8 173.9 196.9 210.5 218.3]`

In [44]:
print(d.sum(axis = 1))
print(d.sum(axis = 0))

[399.2 562.2]
[161.8 173.9 196.9 210.5 218.3]


Exercise 7:

Task: Make a 3x3 identity matrix called `I3`

Desired output: 

`[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]`

In [46]:
I3 = np.eye(3)
I3

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

Exercise 8:

Task: Make a 10x10 array of values 1 to 100. Call it X

Desired output: 

`[[  1   2   3   4   5   6   7   8   9  10]
 [ 11  12  13  14  15  16  17  18  19  20]
 [ 21  22  23  24  25  26  27  28  29  30]
 [ 31  32  33  34  35  36  37  38  39  40]
 [ 41  42  43  44  45  46  47  48  49  50]
 [ 51  52  53  54  55  56  57  58  59  60]
 [ 61  62  63  64  65  66  67  68  69  70]
 [ 71  72  73  74  75  76  77  78  79  80]
 [ 81  82  83  84  85  86  87  88  89  90]
 [ 91  92  93  94  95  96  97  98  99 100]]`

In [49]:
X = np.arange(1,101).reshape(10,10)
X

array([[  1,   2,   3,   4,   5,   6,   7,   8,   9,  10],
       [ 11,  12,  13,  14,  15,  16,  17,  18,  19,  20],
       [ 21,  22,  23,  24,  25,  26,  27,  28,  29,  30],
       [ 31,  32,  33,  34,  35,  36,  37,  38,  39,  40],
       [ 41,  42,  43,  44,  45,  46,  47,  48,  49,  50],
       [ 51,  52,  53,  54,  55,  56,  57,  58,  59,  60],
       [ 61,  62,  63,  64,  65,  66,  67,  68,  69,  70],
       [ 71,  72,  73,  74,  75,  76,  77,  78,  79,  80],
       [ 81,  82,  83,  84,  85,  86,  87,  88,  89,  90],
       [ 91,  92,  93,  94,  95,  96,  97,  98,  99, 100]])

Exercise 9:

Task: Make a copy of X, call it Y (1 line). Replace all values in Y that are not squares with 0 (1 line). see `numpy.isin()`

Desired output: 

`[[  1   0   0   4   0   0   0   0   9   0]
 [  0   0   0   0   0  16   0   0   0   0]
 [  0   0   0   0  25   0   0   0   0   0]
 [  0   0   0   0   0  36   0   0   0   0]
 [  0   0   0   0   0   0   0   0  49   0]
 [  0   0   0   0   0   0   0   0   0   0]
 [  0   0   0  64   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0   0]
 [ 81   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0 100]]`

In [56]:
Y = X.copy()
Y[~np.isin(Y, np.arange(1,11)**2)] = 0
Y

array([[  1,   0,   0,   4,   0,   0,   0,   0,   9,   0],
       [  0,   0,   0,   0,   0,  16,   0,   0,   0,   0],
       [  0,   0,   0,   0,  25,   0,   0,   0,   0,   0],
       [  0,   0,   0,   0,   0,  36,   0,   0,   0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,  49,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0],
       [  0,   0,   0,  64,   0,   0,   0,   0,   0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0],
       [ 81,   0,   0,   0,   0,   0,   0,   0,   0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0, 100]])

Exercise 10:

Task: Tile a 2x2 diagonal matrix of integers to make a checkerboard pattern. Call the matrix `check`

Desired output: 

`[[1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]]`

In [61]:
check = np.tile(np.eye(2), reps=(4,4)).astype(int)
check

array([[1, 0, 1, 0, 1, 0, 1, 0],
       [0, 1, 0, 1, 0, 1, 0, 1],
       [1, 0, 1, 0, 1, 0, 1, 0],
       [0, 1, 0, 1, 0, 1, 0, 1],
       [1, 0, 1, 0, 1, 0, 1, 0],
       [0, 1, 0, 1, 0, 1, 0, 1],
       [1, 0, 1, 0, 1, 0, 1, 0],
       [0, 1, 0, 1, 0, 1, 0, 1]])

Exercise 11:

Task: convert the values in `f_temp` from Farenheit to celsius. The conversion is subtract 32, then multiply by 5/9. Round to 1 decimal place.

Desired output: 

`[[21.1 21.7 22.2 22.8 23.3 23.9]
 [24.4 25.  25.6 26.1 26.7 27.2]
 [27.8 28.3 28.9 29.4 30.  30.6]
 [31.1 31.7 32.2 32.8 33.3 33.9]
 [34.4 35.  35.6 36.1 36.7 37.2]]
`

In [4]:
# do not modify
f_temp  = np.arange(70,100).reshape((5,6))

In [64]:
np.round((f_temp - 32)*5/9, 1)

array([[21.1, 21.7, 22.2, 22.8, 23.3, 23.9],
       [24.4, 25. , 25.6, 26.1, 26.7, 27.2],
       [27.8, 28.3, 28.9, 29.4, 30. , 30.6],
       [31.1, 31.7, 32.2, 32.8, 33.3, 33.9],
       [34.4, 35. , 35.6, 36.1, 36.7, 37.2]])

Exercise 12:

Task: Convert values in the matrix `x` into z-scores by column, call it matrix `z`. That is: each column should have a mean of 0 and std of 1. (subtract the column mean, and divide by the column std). (not required, but see if you can do this in one line)

Print the column means and column std to show that they have been standardized.

Desired output: 

`[[-1.09996745 -0.47901666  0.8816739 ]
 [ 0.9495002   1.18844641  0.11324252]
 [-1.04808219  1.64320907  0.27501755]
 [ 1.23486912  0.40019114  0.84123014]
 [ 1.23486912 -0.78219177 -0.45297008]
 [-0.68488539 -0.75187426  0.5985676 ]
 [ 0.19716398 -0.72155675 -1.46406399]
 [ 1.10515598  1.40066898 -1.74717029]
 [-0.60705751 -1.08536687 -0.57430135]
 [-1.28156585 -0.81250928  1.52877401]]`
 
 `[ 6.66133815e-17  1.11022302e-16 -1.11022302e-16]
[1. 1. 1.]`

In [66]:
# do not modify
np.random.seed(100)
x = np.random.randint(100,size = 30).reshape(10,3)

In [73]:
z = (x - x.mean(axis = 0)) / x.std(axis = 0)
print(z)
z.mean(axis = 0)
z.std(axis = 0)

[[-1.09996745 -0.47901666  0.8816739 ]
 [ 0.9495002   1.18844641  0.11324252]
 [-1.04808219  1.64320907  0.27501755]
 [ 1.23486912  0.40019114  0.84123014]
 [ 1.23486912 -0.78219177 -0.45297008]
 [-0.68488539 -0.75187426  0.5985676 ]
 [ 0.19716398 -0.72155675 -1.46406399]
 [ 1.10515598  1.40066898 -1.74717029]
 [-0.60705751 -1.08536687 -0.57430135]
 [-1.28156585 -0.81250928  1.52877401]]


array([1., 1., 1.])

Exercise 13:

Task: Convert values in the matrix `x` into scaled values from 0 to 10. That is take each column and scale values linearly so that the largest value is 10, and the smallest value in the column is 0. Round results to 2 decimal places. Call the result `y`

(Not required, but see if you can do the calculations in one line.)

Desired output: 

`[[ 0.72  2.22  8.02]
 [ 8.87  8.33  5.68]
 [ 0.93 10.    6.17]
 [10.    5.44  7.9 ]
 [10.    1.11  3.95]
 [ 2.37  1.22  7.16]
 [ 5.88  1.33  0.86]
 [ 9.48  9.11  0.  ]
 [ 2.68  0.    3.58]
 [ 0.    1.   10.  ]]`

In [80]:
# do not modify
np.random.seed(100)
x = np.random.randint(100,size = 30).reshape(10,3)

In [83]:
y = np.round((x - x.min(axis=0)) / (x - x.min(axis=0)).max(axis=0) * 10, decimals=2)
y

array([[ 0.72,  2.22,  8.02],
       [ 8.87,  8.33,  5.68],
       [ 0.93, 10.  ,  6.17],
       [10.  ,  5.44,  7.9 ],
       [10.  ,  1.11,  3.95],
       [ 2.37,  1.22,  7.16],
       [ 5.88,  1.33,  0.86],
       [ 9.48,  9.11,  0.  ],
       [ 2.68,  0.  ,  3.58],
       [ 0.  ,  1.  , 10.  ]])

Exercise 14:

Task: Replace all NaN values in the matrix `x` with 0.

Desired output: 

`[[ 8. 24. 67.]
 [87. 79.  0.]
 [10.  0. 52.]
 [ 0. 53. 66.]]`

In [84]:
# do not modify
np.random.seed(100)
x = np.random.randint(100,size = 12).reshape(4,3).astype('float')
row = np.array([1, 2, 3])
col = np.array([2, 1, 0])
x[row, col] = np.nan

In [87]:
x[np.isnan(x)] = 0
x

array([[ 8., 24., 67.],
       [87., 79.,  0.],
       [10.,  0., 52.],
       [ 0., 53., 66.]])

Exercise 15:

Task: Use a list comprehension to turn the string into a list of letters.

Desired output: 

`['a' 'b' 'c' 'd' 'e' 'f' 'g' 'h' 'i' 'j' 'k' 'l' 'm' 'n' 'o' 'p' 'q' 'r'
 's' 't' 'u' 'v' 'w' 'x' 'y' 'z']`

In [89]:
# do not modify
letters = 'abcdefghijklmnopqrstuvwxyz'

In [95]:
np.array([let for let in letters])

array(['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm',
       'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z'],
      dtype='<U1')

Exercise 16:

Task: Empirically test out the central limit theorem.

- Create an empty list `sample_means_n16`
- Create a loop that will perform 1000 iterations
- Within each loop draw a random sample of 16 values from the normal distribution with mean 100 and sd 20
- Calculate the mean of your random sample and append the new value to `sample_means_n16`
- See that the mean of the sample means is close to the population mean and that the sd of the sample means is approx sigma/sqrt(n)

- see if you can make the loop with only one line of code

Desired output: (you answers may vary)

`100.07429950480352
5.0548725387880635`

In [107]:
np.random.seed(5)
sample_means_n16 = []

In [111]:
for _ in range(1000):
    samp = np.random.normal(100,20,16)
    sample_means_n16.append(samp.mean())
print(np.std(sample_means_n16))
np.mean(sample_means_n16)

5.055861231020289


100.09639496082309

In [117]:
sample_means_n16 = np.random.normal(100,20, (16,1000)).mean(axis = 0)
print(sample_means_n16.std())
sample_means_n16.mean()

5.059407855607151


99.77769219772085

Exercise 17:

Task: Do the same thing, but with a list comprehension

- Using a list comprehension, create `sample_means_n16`
- The list will contain 1000 sample means. Each value mean is the sample mean of a random sample of 16 values from the normal distribution with mean 100 and sd 20
- See that the mean of the sample means is close to the population mean and that the sd of the sample means is approx sigma/sqrt(n)

Desired output: (you answers may vary)

`100.07429950480352
5.0548725387880635`

In [129]:
np.random.seed(5)
sample_means_n16 = np.array([np.random.normal(100,20,16).mean() for _ in range(1000)])

In [130]:
print(sample_means_n16.std())
sample_means_n16.mean()

5.0548725387880635


100.07429950480352

Exercise 18:

Task: Use a list comprehension. Go through the provided text. In the list, store how many words are in each sentence. hint: find sentences by splitting on the period punctuation, and then find words by splitting on space.

Desired output:

`[9, 8, 11, 12, 4, 8, 26]`

In [132]:
# do not modify
text = '''California is a state in the western United 
States. It is the third largest state in size. It is 
the state with the most people living in it. Largest 
cities are Los Angeles, San Diego, San Jose, and San 
Francisco. The capital is Sacramento. It became a 
state on September 9, 1850. It is bordered by Arizona 
to the southeast, Oregon to the north, Nevada to the 
east and the Mexican state of Baja California to the 
south.'''

In [164]:
sentences = [sent.strip() for sent in text.split('.') if sent]
[len(words.split()) for words in sentences]

[9, 8, 11, 12, 4, 8, 26]